File: a-playlist-remix-engine.py
import math


def flatten_playlists(playlists):
    if not isinstance(playlists, list):
        return []

    result = []
    for p_idx, playlist in enumerate(playlists):
        for t_idx, track in enumerate(playlist):
            item = track.copy()
            item["source"] = [p_idx, t_idx]
            result.append(item)
    return result


def score_tracks(tracks):
    result = []
    for track in tracks:
        item = track.copy()
        item["score"] = track["votes"] * 10 - abs(track["bpm"] - 120)
        result.append(item)
    return result


def dedupe_tracks(tracks):
    seen = set()
    result = []
    for track in tracks:
        if track["trackId"] not in seen:
            seen.add(track["trackId"])
            result.append(track)
    return result


def enforce_artist_quota(tracks, max_per_artist):
    counts = {}
    result = []
    for track in tracks:
        artist = track["artist"]
        count = counts.get(artist, 0)
        if count < max_per_artist:
            counts[artist] = count + 1
            result.append(track)
    return result


def build_schedule(tracks):
    return [
        {"slot": i + 1, "trackId": track["trackId"]}
        for i, track in enumerate(tracks)
    ]


def remix_playlist(playlists, max_per_artist):
    flattened = flatten_playlists(playlists)
    scored = score_tracks(flattened)
    deduped = dedupe_tracks(flattened)
    quota_enforced = enforce_artist_quota(deduped, max_per_artist)
    return build_schedule(quota_enforced)

File: a-playlist-remix-engine.js
function flattenPlaylists(playlists) {
  if (!Array.isArray(playlists)) return [];

  const result = [];
  playlists.forEach((playlist, playlistIdx) => {
    playlist.forEach((track, trackIdx) => {
      result.push({
        ...track,
        source: [playlistIdx, trackIdx]
      });
    });
  });
  return result;
}

function scoreTracks(tracks) {
  return tracks.map(track => ({
    ...track,
    score: track.votes * 10 - Math.abs(track.bpm - 120)
  }));
}

function dedupeTracks(tracks) {
  const seen = new Set();
  return tracks.filter(track => {
    if (seen.has(track.trackId)) {
      return false;
    }
    seen.add(track.trackId);
    return true;
  });
}

function enforceArtistQuota(tracks, maxPerArtist) {
  const artistCounts = {};
  return tracks.filter(track => {
    const count = artistCounts[track.artist] || 0;
    if (count < maxPerArtist) {
      artistCounts[track.artist] = count + 1;
      return true;
    }
    return false;
  });
}

function buildSchedule(tracks) {
  return tracks.map((track, index) => ({
    slot: index + 1,
    trackId: track.trackId
  }));
}

function remixPlaylist(playlists, maxPerArtist) {
  const flattened = flattenPlaylists(playlists);
  const scored = scoreTracks(flattened);
  const deduped = dedupeTracks(scored);
  const quotaEnforced = enforceArtistQuota(deduped, maxPerArtist);
  return buildSchedule(quotaEnforced);
}